pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -57,6 +58,12 @@
 @Slf4j
 public class PersistentAcknowledgmentsGroupingTracker implements AcknowledgmentsGroupingTracker {
 
+    private static final TimedCompletableFuture<Void> COMPLETED_FUTURE = new TimedCompletableFuture<>();
```

Comment:
please NEVER use a static variable for a CompletableFuture.
The behaviour is really unpredictable.
you cannot ever cache a CF

Reply:
Sure. But could you explain more about why a static CompletableFuture cannot be used?

Reply:
Because CompletableFuture is stateful and it has internal queues of pending threads that wait for things to happen.
You usually create chains of other CompletionStage instances created by the initial CF and this easily turns into unpredictable behaviour and also memory leaks.